### PR TITLE
Explain why we check hashes before signatures

### LINKED ISSUE #138: Explain why checking some hashes before signatures is ok?
From this [PR on php-tuf](https://github.com/php-tuf/php-tuf/pull/88#issuecomment-737045948), I learned that it is not clear why the spec advises on checking snapshot/targets hashes from timestamp/snapshot before checking signatures. It looks unsafe until it is recalled that the hashes from previously verified snapshot/targets. Recently while independently implementing the spec, even I had some concern until I understood this. Should we clarify this?

HT @tedbow @davidstrauss

### PR DESCRIPTION
**Pull request by trishankatdatadog**
An attempt to fix #138 

Does this look good @tedbow?

Signed-off-by: Trishank Karthik Kuppusamy <trishank.kuppusamy@datadoghq.com>

### CODE REVIEW

**Review by lukpueh — COMMENTED** [review]
Thanks for clarifying this, @trishankatdatadog. I wonder if we can make it more concise.

>It is safe to check the hashes before the signatures, because the hashes comes from the timestamp role, which we have already verified in the previous step

feels a bit redundant to 
>  hashes ... listed in the trusted timestamp metadata

Maybe it's enough to just expand what's already there, e.g.  
> hashes ... listed in the above/previously/already verified and thus trusted timestamp metadata

?


On the other hand, it seems worthwhile to elaborate on the 
> quick way to reject bad metadata

What about something along the lines of

> preliminary integrity check before performing a more expensive signature verification

?

Besides, should this be part of the spec or is it rather a contender for secondary literature (https://github.com/theupdateframework/specification/issues/91)?

**Comment by trishankatdatadog** [comment]
Sounds good, I will address your comments. In the meantime, I think we should add this to the spec, because that's where the question came up for php-tuf while implementing, but I'll let @tedbow decide.

**Comment by joshuagl** [comment]
I agree that the spec is a good place for this, at least until we have the secondary literature. There are other parts of the spec where we explain why things are they way they are, such as the recently introduced fixed update start time.

**Comment by trishankatdatadog** [comment]
@joshuagl @mnm678 PTAL 👀

File: tuf-spec.md [hunk 1]
```diff
@@ -1448,9 +1450,11 @@ it in the next step.
 2. **Check against snapshot role's targets hash**. The hashes
   of the new targets metadata file MUST match the hashes, if any, listed in the
   trusted snapshot metadata.  This is done, in part, to prevent a mix-and-match
-  attack by man-in-the-middle attackers.  If the new targets metadata file does
-  not match, discard the new target metadata, abort the update cycle, and
-  report the failure.
+  attack by man-in-the-middle attackers. It is safe to check the hashes before
+  the signatures, because the hashes comes from the snapshot role, which we have
```
Comment (joshuagl) [inline tuf-spec.md:1454]:
```suggestion
  the signatures, because the hashes come from the snapshot role, which we have
```

File: tuf-spec.md [hunk 2]
```diff
@@ -1399,9 +1399,11 @@ it in the next step.
 2. **Check against timestamp role's snapshot hash**. The hashes
   of the new snapshot metadata file MUST match the hashes, if any, listed in
   the trusted timestamp metadata.  This is done, in part, to prevent a
-  mix-and-match attack by man-in-the-middle attackers.  If the hashes do not
-  match, discard the new snapshot metadata, abort the update cycle, and report
-  the failure.
+  mix-and-match attack by man-in-the-middle attackers. It is safe to check the
+  hashes before the signatures, because the hashes comes from the timestamp
```
Comment (joshuagl) [inline tuf-spec.md:1403]:
```suggestion
  hashes before the signatures, because the hashes come from the timestamp
```

**Review by joshuagl — APPROVED** [review]
LGTM, thanks Trishank

**Review by joshuagl — APPROVED** [review]

**Review by mnm678 — APPROVED** [review]

**Pull request merged**